editor/src/java/com/defold/util/WeakInterner.java

```diff
@@ -0,0 +1,286 @@
+     * @return The canonical representation of the supplied value.
+     */
+    public T intern(final T value) {
+        if (value == null) {
+            throw new IllegalArgumentException("The interned value cannot be null.");
+        }
+
+        // First attempt to get the existing value without locking. If we don't
+        // find a match, lock access from other threads while adding it.
+        final int hashValue = getHashValue(value);
+        final Entry<T>[] hashTable = getHashTable();
+        final int bucketIndex = getBucketIndex(hashValue, hashTable.length);
+        final Entry<T> firstEntryInBucket = hashTable[bucketIndex];
+        final T existingValue = findExistingValue(firstEntryInBucket, value, hashValue);
+        return existingValue != null ? existingValue : internSynchronized(value, hashValue);
+    }
+
+    private synchronized T internSynchronized(final T value, final int hashValue) {
+        // Another thread might have modified the hash table while we were
+        // waiting for it to release the lock. Thus, we must get the hash table
+        // again and try to find an existing value once more before interning it
+        // ourselves.
+        final Entry<T>[] hashTable = getHashTable();
+        final int bucketIndex = getBucketIndex(hashValue, hashTable.length);
+        final Entry<T> firstEntryInBucket = hashTable[bucketIndex];
+        final T existingValue = findExistingValue(firstEntryInBucket, value, hashValue);
+
+        if (existingValue != null) {
+            // Another thread interned the value while we were waiting.
+            return existingValue;
+        }
+
+        // We get to intern the value.
+        hashTable[bucketIndex] = new Entry<>(value, staleEntriesQueue, hashValue, firstEntryInBucket);
+
+        if (++count >= growthThreshold) {
+            rehash(hashTable.length * 2);
+        }
+
+        return value;
+    }
+
+    private static <T> T findExistingValue(final Entry<T> firstEntryInBucket, final T value, final int hashValue) {
+        for (Entry<T> existingEntry = firstEntryInBucket; existingEntry != null; existingEntry = existingEntry.nextEntry) {
+            if (hashValue != existingEntry.hashValue) {
+                continue;
+            }
+
+            if (existingEntry.refersTo(value)) {
+                return value;
+            }
+
+            final T existingValue = existingEntry.get();
+
+            if (value.equals(existingValue)) {
+                return existingValue;
+            }
+        }
+
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> Entry<T>[] makeHashTable(final int capacity) {
+        return (Entry<T>[]) new Entry<?>[capacity];
+    }
+
+    private Entry<T>[] getHashTable() {
+        removeStaleEntries();
+        return hashTable;
+    }
+
+    private void removeStaleEntries() {
+        for (Object staleEntry; (staleEntry = staleEntriesQueue.poll()) != null;) {
```

Comment:
The `poll` method is thread-safe, but we don't have to lock on ourselves until we actually need to remove an entry.